typing/typecore.ml

```diff
@@ -961,7 +963,22 @@ module Constructor = NameChoice (struct
   let type_kind = "variant"
   let get_name cstr = cstr.cstr_name
   let get_type cstr = cstr.cstr_res
-  let get_descrs = fst
+  let extend t env = match t with
+    | Path.Papply _ | Path.Pident _ -> env
+    | Path.Pdot(l,_,_) -> match Env.open_signature Asttypes.Override l env with
+      | Some env -> env
+      | None -> env
+  let spellchecking_candidates env t =
+    let cstrs = fst ( Env.find_type_descrs t env ) in
+    if cstrs = [] then (* search for extension constructors in *)
+      Env.fold_constructors List.cons None env []
+    else cstrs
+  let disambiguation_candidates env t s =
+    let cstrs = fst ( Env.find_type_descrs t env ) in
+    if cstrs = [] then (* extension constructor path *)
+      let env = extend t env in
+      List.map fst ( Env.lookup_all_constructors (Longident.Lident s) env )
+    else cstrs
```

Comment:
Since we have empty sum type now, `cstrs = []` doesn't necessarily mean it's an extension constructor. That being said, this simply means doing extra work in a very uncommon cases, and doesn't compromise soundness, so I guess it's fine. 

You could make this a bit better with the (unexposed) `Env.find_type_full` function I suppose.

Reply:
I think I will start by extending the comment, I believe that optimization for empty sum types can come later.